### PR TITLE
Entypo+ and associated defs, many new docs, colorpicker, updated examples

### PR DESCRIPTION
**Pull request by svenevs**
Definitely squash this, there was a lot of intermediate commits for the entypo stuff that have now been reverted or destroyed.

~~**NOTE**: I'm going through and doing a PR review right now.  Don't review until this is gone since this is a huge update and most of it is just noise from doc comments.~~

1. It would be kind of tedious to separate this out into multiple PRs, but it can be done if you prefer.  Things like docs links got added in a few different locations relating to entypo.

2. See annotations in PR review.  Basically, this all boils down to the following:

    - Entypo+ updates
        - Some constants have changed, those are all fixed.  Entypo+ file documentation enumerates everything, as well as there are CPP and python examples.
        - Python binding code changes slightly to allow auto-generation of Entypo+ constants without having to incorporate GLFW stuff in the gen code.
        - **Important**: icon sizes have changed.  See [relevant change to common.h](https://github.com/wjakob/nanogui/pull/260#discussion_r139310396).  Here is a full index of where that applies:
            - [first scale change](https://github.com/wjakob/nanogui/pull/260#discussion_r139310615)
            - [second scale change](https://github.com/wjakob/nanogui/pull/260#discussion_r139310622)
            - [third scale change](https://github.com/wjakob/nanogui/pull/260#discussion_r139310634)
            - [fourth scale change](https://github.com/wjakob/nanogui/pull/260#discussion_r139310724)
            - [fifth scale change](https://github.com/wjakob/nanogui/pull/260#discussion_r139310729)
            - [sixth scale change](https://github.com/wjakob/nanogui/pull/260#discussion_r139310735)
            - [seventh scale change](https://github.com/wjakob/nanogui/pull/260#discussion_r139310748)
            - Basically, the ones using `ih` should be good.  But a couple were not scaled the same as previous code.
    - ColorPicker has been modified
        - I believe the new functionality makes sense.  These changes came about from writing the docs and not understanding how it worked.  The callback gets executed too frequently.  Since it's used in the form helper, I decided it needed to change ;)
    - Sad face.  Docs required getting rid of `using` for eigen vector types :/

3. I'm going to split off onto another branch now for the remaining docs updates.  I made a serializer example, but that and the changes made for the docs will be in their own self-contained PR because they aren't exactly trivial changes.

The rest is just docs updates really.  Highlight of fake docs build [entypo.h](http://fake-nanogui.readthedocs.io/en/latest/api/file_nanogui_entypo.h.html#file-nanogui-entypo-h) :)

P.S. In future merge, you may need to **wipe** the environment for RTD to build it right because of the change in `requirements.txt` and more specifically if `exhale.pyc` is still floating around it will complain about no `setup(app)` "are you sure it is a sphinx extension" because it wasn't.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -433,7 +435,8 @@ if (NANOGUI_BUILD_PYTHON)
   include_directories("ext/pybind11/include" ${PYTHON_INCLUDE_DIR})
   add_library(nanogui-python-obj OBJECT
     python/main.cpp
-    python/constants.cpp
+    python/constants_glfw.cpp
```
Comment (svenevs) [inline CMakeLists.txt]: `python/constants.cpp` got split into two files.

1. `constants_glfw.cpp` has same as before, just in a different file.
2. `constants_entypo.cpp` is generated (which is why they got split)

File: docs/conf.py [hunk 2]
```diff
@@ -32,14 +32,85 @@
...
+    "containmentFolder":     "./api",
+    "rootFileName":          "library_root.rst",
+    "rootFileTitle":         "Library API",
+    "doxygenStripFromPath":  "../include",
+    ############################################################################
+    # Suggested optional arguments.                                            #
+    ############################################################################
+    "createTreeView":        True,
+    "exhaleExecutesDoxygen": True,
+    "exhaleDoxygenStdin": textwrap.dedent('''
```
Comment (svenevs) [inline docs/conf.py]: For the most part, all the other Doxygen stuff from the original `conf.py` are in the Exhale defaults.  We only need `INPUT` and `PREDEFINED` for the namespaces.  I still have no idea why Doxygen can't handle those.

File: docs/conf.py [hunk 3]
```diff
@@ -32,14 +32,85 @@
...
+        PREDEFINED            += DOXYGEN_DOCUMENTATION_BUILD
+        PREDEFINED            += NANOGUI_EXPORT
+    '''),
+    ############################################################################
+    # HTML Theme specific configurations.                                      #
+    ############################################################################
+    # Fix broken Sphinx RTD Theme 'Edit on GitHub' links
+    # Search for 'Edit on GitHub' on the FAQ:
+    #     http://exhale.readthedocs.io/en/latest/faq.html
+    "pageLevelConfigMeta": ":github_url: https://github.com/wjakob/nanogui",
```
Comment (svenevs) [inline docs/conf.py]: This is the (temporary, but not going to be fixed any time soon) hack to get the generated pages to link back to a real URL.

File: docs/conf.py [hunk 4]
```diff
@@ -303,102 +366,7 @@
...
-        "createTreeView"             : True,
-        "afterTitleDescription"      : libraryDescription,
-        "doxygenStripFromPath"       : ".."
-    }
-
-    # import the exhale module from the current directory and generate the api
-    from exhale import generate
-    generate(exhaleArgs)
+    # this looks in html_static_path (_static directory as set above)
+    app.add_stylesheet("entypo.css")
```
Comment (svenevs) [inline docs/conf.py]: Note: `fontcustom` generated `entypo.css` will **NOT** conflict with sphinx theme because all CSS targets are `entypo-icon-X` rather than just `icon-X`.  AKA even though this is linked to on every page, it doesn't cause issues.

File: docs/examples.rst [hunk 5]
```diff
@@ -58,6 +60,24 @@ arbitrary sequence of OpenGL commands into a NanoGUI widget.
 - `Example 4 in C++ <https://github.com/wjakob/nanogui/blob/master/src/example4.cpp>`_
 - `Example 4 in Python <https://github.com/wjakob/nanogui/blob/master/python/example4.py>`_
 
+.. _nanogui_example_icons:
+
+Example Icons
```
Comment (svenevs) [inline docs/examples.rst]: These were added for testing, but I don't see the harm in having them.  The icons will show up slightly differently than the webpage because of the default dark background of NanoGUI.  Can remove if undesired.

File: docs/index.rst [hunk 6]
```diff
@@ -64,7 +64,7 @@ Contents
    usage
    compilation
    examples
-   generated_api/library_root
+   api/library_root
```
Comment (svenevs) [inline docs/index.rst]: This change means that people who have setup non intersphinx links will now direct to invalid URLs.  But the new url makes more sense, as well as `exhale_` is removed from the beginning of every filename as it now requires a subdirectory in the sphinx source dir.  `exhale_` prefix was because it wasn't a strict requirement before.

File: docs/requirements.txt [hunk 7]
```diff
@@ -1,2 +1 @@
-breathe
-# git+git://github.com/michaeljones/breathe@v4.3.1#egg=breathe
+exhale
```
Comment (svenevs) [inline docs/requirements.txt]: To build locally, you'll need to `pip3 install exhale`.  NOTE: py3 is noticeably faster than py2 ;)

File: include/nanogui/colorpicker.h [hunk 8]
```diff
@@ -21,24 +22,84 @@ NAMESPACE_BEGIN(nanogui)
...
-    /// Get the current color
+    /// Get the current Color selected for this ColorPicker.
     Color color() const;
-    /// Set the current color
+
+    /// Set the current Color selected for this ColorPicker.
     void setColor(const Color& color);
+
+    /// The current caption of the \ref nanogui::ColorPicker::mPickButton.
+    const std::string &pickButtonCaption() { return mPickButton->caption(); }
```
Comment (svenevs) [inline include/nanogui/colorpicker.h]: Added pass-through methods for setting button captions if desired.

File: include/nanogui/colorpicker.h [hunk 9]
```diff
@@ -21,24 +22,84 @@ NAMESPACE_BEGIN(nanogui)
...
     Button *mPickButton;
+
+    /**
+     * Remains the Color of the active color selection, until the user picks a
+     * new Color on the ColorWheel **and** selects the
+     * \ref nanogui::ColorPicker::mPickButton.  The default value for the
+     * caption of this Button is ``"Reset"``.  You can change it using
+     * \ref nanogui::ColorPicker::setResetButtonCaption if you need.
+     */
+    Button *mResetButton;
```
Comment (svenevs) [inline include/nanogui/colorpicker.h]: This is new.  The changed functionality of ColorPicker necessitates this.  See comments in cpp file.

File: include/nanogui/common.h [hunk 10]
```diff
@@ -109,39 +111,47 @@ struct GLFWcursor;
...
-using Eigen::Vector4f;
-using Eigen::Vector2i;
-using Eigen::Vector3i;
-using Eigen::Vector4i;
-using Eigen::Matrix3f;
-using Eigen::Matrix4f;
-using Eigen::VectorXf;
-using Eigen::MatrixXf;
+/// Type alias to allow ``Eigen::Vector2f`` to be used as ``nanogui::Vector2f``.
+typedef Eigen::Vector2f Vector2f;
```
Comment (svenevs) [inline include/nanogui/common.h]: Breathe chokes on raw `using` in a namespace (it does seem to do ok with them in a class).  Without explicit typedefs here, the fact that `nanogui::Vector3f` etc are available never show up in the docs.

This can easily be undone, at the expense of no explicit documentation of the vector types.

File: include/nanogui/common.h [hunk 11]
```diff
@@ -494,7 +504,13 @@ extern NANOGUI_EXPORT std::vector<std::pair<int, std::string>>
 
 /// Convenience function for instanting a PNG icon from the application's data segment (via bin2c)
 #define nvgImageIcon(ctx, name) nanogui::__nanogui_get_image(ctx, #name, name##_png, name##_png_size)
+
 /// Helper function used by nvgImageIcon
 extern NANOGUI_EXPORT int __nanogui_get_image(NVGcontext *ctx, const std::string &name, uint8_t *data, uint32_t size);
 
+namespace detail {
+    /// Used for scaling icons (primarily Entypo) in widget draw calls.
+    static constexpr float __icon_scale = 0.77f;
```
Comment (svenevs) [inline include/nanogui/common.h]: The new Entypo+ font is presumably bigger.  See annotations in various CPP files for how this is used.

File: include/nanogui/glcanvas.h [hunk 12]
```diff
@@ -21,47 +25,67 @@ NAMESPACE_BEGIN(nanogui)
...
+ *        contributed by Jan Winkler.
  *
  * Canvas widget that can be used to display arbitrary OpenGL content. This is
  * useful to display and manipulate 3D objects as part of an interactive
  * application. The implementation uses scissoring to ensure that rendered
  * objects don't spill into neighboring widgets.
  *
- * Usage: override `drawGL` in subclasses to provide custom drawing code.
+ * \rst
+ * **Usage**
```
Comment (svenevs) [inline include/nanogui/glcanvas.h]: For the most part, all of @fairlight1337 's comments stayed the same.  I just linked back to the example, and put some `.`'s at the end of doc comments because I've been obsessively doing that to everything.  AKA these doc changes are just for consistency, but I also added that he contributed the widget :)

File: include/nanogui/opengl.h [hunk 13]
```diff
@@ -15,6 +15,10 @@
 
 #include <nanogui/common.h>
 
+// The Entypo PLUS and MINUS icons are 43 and 45 respectively, which needs to
+// be checked manually.
+#include <nanogui/entypo.h>
```
Comment (svenevs) [inline include/nanogui/opengl.h]: Ooops.  Need to remove this, will happen in next commit on this branch.  This was the weird font encoding issue, was comparing to those two icon values explicitly in the helper functions.

File: include/nanogui/textbox.h [hunk 14]
```diff
@@ -66,8 +66,10 @@ class NANOGUI_EXPORT TextBox : public Widget {
     /// Set the \ref Theme used to draw this widget
     virtual void setTheme(Theme *theme) override;
 
-    /// Set the change callback
+    /// The callback to execute when the value of this TextBox has changed.
```
Comment (svenevs) [inline include/nanogui/textbox.h]: The rest of `textbox.h` is still a TODO.  I only updated things to make docs link back and forth.  Same happened in `widget.h`.

File: python/constants.cpp [hunk 15]
```diff
@@ -1,107 +0,0 @@
-#ifdef NANOGUI_PYTHON
```
Comment (svenevs) [inline python/constants.cpp]: As discussed above, this file is now gone in favor of two (one for GLFW, one for Entypo)

File: python/eigen.cpp [hunk 16]
```diff
@@ -16,6 +16,8 @@ void register_eigen(py::module &m) {
                       [](Color &c, float v) { c.g() = v; }, D(Color, g))
         .def_property("b", [](const Color &c) { return c.b(); },
                       [](Color &c, float v) { c.b() = v; }, D(Color, b))
+        .def_property("w", [](const Color &c) { return c.w(); },
```
Comment (svenevs) [inline python/eigen.cpp]: This one needs to be bound with an explicit docstring because `make mkdoc` will not get the inherited Eigen docs (which is where `Vector4f::w()` comes from).  This is used in the new examples.

File: python/example1.py [hunk 17]
```diff
@@ -278,32 +278,19 @@ def cb(value):
...
-        popup.setLayout(GroupLayout())
-
-        colorwheel = ColorWheel(popup)
-        colorwheel.setColor(popupBtn.backgroundColor())
-
-        colorBtn = Button(popup, "Pick")
-        colorBtn.setFixedSize((100, 25))
-        c = colorwheel.color()
-        colorBtn.setBackgroundColor(c)
+        Label(window, "Color picker :", "sans-bold");
```
Comment (svenevs) [inline python/example1.py]: Example 1 in python; Entypo icon fixes and using ColorPicker directly.

File: python/example2.py [hunk 18]
```diff
@@ -55,7 +55,16 @@ def getter():
...
+
+def cp_cb(color):
+    print(
+        "ColorPicker: [{0}, {1}, {2}, {3}]".format(color.r,
+                                                   color.g,
+                                                   color.b,
+                                                   color.w)
+    )
+
+gui.addColorVariable("Color", *make_accessors("colvar")).setCallback(cp_cb)
```
Comment (svenevs) [inline python/example2.py]: FormHelper uses ColorPicker, add callback to demonstrate its new functionality.

File: python/exampleIcons.py [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+
+    # vscroll should only have *ONE* child. this is what `wrapper` is for
+    wrapper = Widget(vscroll)
+    wrapper.setFixedSize((width, height))
+    wrapper.setLayout(GridLayout())  # defaults: 2 columns
+
+    # NOTE: don't __dict__ crawl in real code!
+    # this is just because it's more convenient to do this for enumerating all
+    # of the icons -- see cpp example for alternative...
+    for key in entypo.__dict__.keys():
```
Comment (svenevs) [inline python/exampleIcons.py]: I'd rather not do what was done in C++, but since this is an example, maybe this is bad?

File: python/main.cpp [hunk 20]
```diff
@@ -37,7 +37,8 @@ namespace {
 }
 #endif
 
-extern void register_constants(py::module &m);
+extern void register_constants_glfw(py::module &m);
```
Comment (svenevs) [inline python/main.cpp]: The corresponding updates to the python bindings.

File: src/button.cpp [hunk 21]
```diff
@@ -31,7 +31,7 @@ Vector2i Button::preferredSize(NVGcontext *ctx) const {
 
     if (mIcon) {
         if (nvgIsFontIcon(mIcon)) {
-            ih *= 1.5f;
+            ih *= detail::__icon_scale;
```
Comment (svenevs) [inline src/button.cpp]: First icon scaling change.

File: src/button.cpp [hunk 22]
```diff
@@ -169,7 +169,7 @@ void Button::draw(NVGcontext *ctx) {
 
         float iw, ih = fontSize;
         if (nvgIsFontIcon(mIcon)) {
-            ih *= 1.5f;
+            ih *= detail::__icon_scale;
```
Comment (svenevs) [inline src/button.cpp]: Second icon scaling change.

File: src/checkbox.cpp [hunk 23]
```diff
@@ -78,7 +78,7 @@ void CheckBox::draw(NVGcontext *ctx) {
     nvgFill(ctx);
 
     if (mChecked) {
-        nvgFontSize(ctx, 1.8 * mSize.y());
+        nvgFontSize(ctx, detail::__icon_scale * mSize.y());
```
Comment (svenevs) [inline src/checkbox.cpp]: Third icon scaling change.  I got lazy and used one globally, this one may need to change.

File: src/colorpicker.cpp [hunk 24]
```diff
@@ -22,26 +22,53 @@ ColorPicker::ColorPicker(Widget *parent, const Color& color) : PopupButton(paren
...
+    mResetButton->setBackgroundColor(color);
+    mResetButton->setTextColor(color.contrastingColor());
+    mResetButton->setFixedSize(Vector2i(100, 20));
 
     PopupButton::setChangeCallback([&](bool) {
-        setColor(backgroundColor());
-        mCallback(backgroundColor());
+        if (this->mPickButton->pushed()) {
+            setColor(backgroundColor());
+            mCallback(backgroundColor());
```
Comment (svenevs) [inline src/colorpicker.cpp]: Callback was getting executed too frequently, e.g. when a user changes the color wheel value but does NOT select "Pick".  Remaining changes in this file relate to this, as well as the reset button.

File: src/example1.cpp [hunk 25]
```diff
@@ -458,31 +459,15 @@ class ExampleApplication : public nanogui::Screen {
...
-        colorwheel->setCallback([colorBtn](const Color &value) {
-            colorBtn->setBackgroundColor(value);
-        });
-
-        colorBtn->setChangeCallback([colorBtn, popupBtn](bool pushed) {
-            if (pushed) {
-                popupBtn->setBackgroundColor(colorBtn->backgroundColor());
-                popupBtn->setPushed(false);
-            }
+        new Label(window, "Color picker :", "sans-bold");
```
Comment (svenevs) [inline src/example1.cpp]: Corresponding example1 updates in CPP.

File: src/example2.cpp [hunk 26]
```diff
@@ -63,7 +63,14 @@ int main(int /* argc */, char ** /* argv */) {
         gui->addGroup("Complex types");
         gui->addVariable("Enumeration", enumval, enabled)
            ->setItems({"Item 1", "Item 2", "Item 3"});
-        gui->addVariable("Color", colval);
+        gui->addVariable("Color", colval)
+           ->setCallback([](const Color &c) {
```
Comment (svenevs) [inline src/example2.cpp]: Example 2 updates in CPP.

File: src/example3.cpp [hunk 27]
```diff
@@ -113,7 +113,14 @@ int main(int /* argc */, char ** /* argv */) {
 
     gui->addGroup("Complex types");
     gui->addVariable("Enumeration", enumval, enabled)->setItems({ "Item 1", "Item 2", "Item 3" });
-    gui->addVariable("Color", colval);
+    gui->addVariable("Color", colval)
```
Comment (svenevs) [inline src/example3.cpp]: Example 3 updates in CPP.  Does not apply to python example 3.

File: src/popupbutton.cpp [hunk 28]
```diff
@@ -45,7 +45,7 @@ void PopupButton::draw(NVGcontext* ctx) {
         NVGcolor textColor =
             mTextColor.w() == 0 ? mTheme->mTextColor : mTextColor;
 
-        nvgFontSize(ctx, (mFontSize < 0 ? mTheme->mButtonFontSize : mFontSize) * 1.5f);
+        nvgFontSize(ctx, (mFontSize < 0 ? mTheme->mButtonFontSize : mFontSize) * detail::__icon_scale);
```
Comment (svenevs) [inline src/popupbutton.cpp]: Fourth icon scale change.  This one used the original `ih *= 0.5` values, but it may need a different scale.

File: src/tabheader.cpp [hunk 29]
```diff
@@ -407,10 +407,10 @@ void TabHeader::drawControls(NVGcontext* ctx) {
 
     // Draw the arrow.
     nvgBeginPath(ctx);
-    auto iconLeft = utf8(ENTYPO_ICON_LEFT_BOLD);
+    auto iconLeft = utf8(ENTYPO_ICON_ARROW_BOLD_LEFT);
     int fontSize = mFontSize == -1 ? mTheme->mButtonFontSize : mFontSize;
     float ih = fontSize;
-    ih *= 1.5f;
+    ih *= detail::__icon_scale;
```
Comment (svenevs) [inline src/tabheader.cpp]: Fifth icon scale.  Should be good.

File: src/tabheader.cpp [hunk 30]
```diff
@@ -429,10 +429,10 @@ void TabHeader::drawControls(NVGcontext* ctx) {
     active = mVisibleEnd != tabCount();
     // Draw the arrow.
     nvgBeginPath(ctx);
-    auto iconRight = utf8(ENTYPO_ICON_RIGHT_BOLD);
+    auto iconRight = utf8(ENTYPO_ICON_ARROW_BOLD_RIGHT);
     fontSize = mFontSize == -1 ? mTheme->mButtonFontSize : mFontSize;
     ih = fontSize;
-    ih *= 1.5f;
+    ih *= detail::__icon_scale;
```
Comment (svenevs) [inline src/tabheader.cpp]: Sixth icon scale.

File: src/textbox.cpp [hunk 31]
```diff
@@ -151,7 +151,7 @@ void TextBox::draw(NVGcontext* ctx) {
         spinArrowsWidth = 14.f;
 
         nvgFontFace(ctx, "icons");
-        nvgFontSize(ctx, ((mFontSize < 0) ? mTheme->mButtonFontSize : mFontSize) * 1.2f);
+        nvgFontSize(ctx, ((mFontSize < 0) ? mTheme->mButtonFontSize : mFontSize) * detail::__icon_scale);
```
Comment (svenevs) [inline src/textbox.cpp]: Seventh icon scale.  This one is not the "default" `1.5`, and may also need a different value?

File: docs/requirements.txt (hunk 7 above)
Comment (wjakob) [reply docs/requirements.txt]: Nice to hear -- many times, it is unfortunately the other way ;).

File: CMakeLists.txt (hunk 1 above)
Comment (wjakob) [reply CMakeLists.txt]: Good call

File: docs/conf.py (hunk 4 above)
Comment (wjakob) [reply docs/conf.py]: Having the icons in the HTML docs is super-nifty. I always dreaded finding the right symbols with boring recompilation cycles.

File: docs/examples.rst (hunk 5 above)
Comment (wjakob) [reply docs/examples.rst]: Neat. Would you mind calling the executable ``example_icons``?

File: include/nanogui/common.h (hunk 10 above)
Comment (wjakob) [reply include/nanogui/common.h]: Typedefs are so C++03 :(. Can we at least do ``using VectorXf = Eigen::VectorXf``, etc?

File: include/nanogui/glcanvas.h (hunk 12 above)
Comment (wjakob) [reply include/nanogui/glcanvas.h]: Sounds good :)

File: python/exampleIcons.py (hunk 19 above)
Comment (wjakob) [reply python/exampleIcons.py]: It seems Pythonic enough to me, definitely preferable to the C++ version.

File: src/checkbox.cpp (hunk 23 above)
Comment (wjakob) [reply src/checkbox.cpp]: I liked the bigger check mark better. Sounds like you'll need an extra factor of ``1.2f``.

File: src/colorpicker.cpp (hunk 24 above)
Comment (wjakob) [reply src/colorpicker.cpp]: One option would be to split that into two callbacks. For instance, the slider widget has a fast callback and a slow one, which may be useful for quick interactive visualizations.

File: src/textbox.cpp (hunk 31 above)
Comment (wjakob) [reply src/textbox.cpp]: Yep, the little scroll handles got bigger. It looked nicer before.

File: include/nanogui/common.h (hunk 11 above)
Comment (wjakob) [reply include/nanogui/common.h]: Sounds fair. I don't think it needs to have such a cryptic name (``__icon_scale``) and be ``constexpr``. The ``Theme`` class would seem like a more natural home to me.

Comment (wjakob) [reply include/nanogui/common.h]: I just realized that ``entypo`` has a CC-BY-SA license, which is a bit more restrictive than the BSD license that is used for the rest of NanoGUI. Unlikely to be an issue to anyone, but it may be worth adding a note somewhere in the docs.

**Comment by wjakob** [comment]
This looks great, I added a few minor comments.

**Comment by wjakob** [comment]
One more thing: somehow the top icons of the tab widget look messed up for me (try ``example1``). Does it happen on your machine as well?

File: docs/conf.py (hunk 4 above)
Comment (svenevs) [reply docs/conf.py]: LOL yeah it was a little tedious to get started, but the CSS generated can be parsed with regular expressions so once that was done it was good

File: docs/examples.rst (hunk 5 above)
Comment (svenevs) [reply docs/examples.rst]: Done

File: include/nanogui/common.h (hunk 11 above)
Comment (svenevs) [reply include/nanogui/common.h]: Should it go on the README?

Comment (svenevs) [reply include/nanogui/common.h]: As for cryptic name, I'm adding it to the theme.  The `__` came from `__nanogui_get_image` above it...but I see now that's because of the `define` above xD

File: src/colorpicker.cpp (hunk 24 above)
Comment (svenevs) [reply src/colorpicker.cpp]: I'm not sure I understand.  If you change the current `master` of NanoGUI example1 to use the `ColorPicker`, it might be clearer why this change is here.  I rewrote `ColorPicker` to match the existing example1 functionality, specifically that the button color does not change until it's been selected.  The reset button was added (can be removed) to allow the color wheel to be reset to the actively selected color.

How does a different callback come in?

**Comment by svenevs** [comment]
Ok icon scaling attempt redone.  I'm a little concerned about positioning, but I think the easiest solution is to avoid `SMALL` icons.  The alternative solution was too confusing for me, but it would be to somehow account for `Widget::icon_scale()` in the width positioning.  I got really turned around trying that though...

If you use this branch and set change `popupbutton.cpp`'s constructor at the top to use `ENTYPO_ICON_CHEVRON_SMALL_RIGHT` and comment out `mIconExtraScale`, Example1 specifically on the Combo Box is wrong.  I added a left popup to make sure nothing went wrong there, and I think the underlying symptom is that `Combo box item 1` is starting to get too big?  Basically, I'm not sure if this is a hidden bug, or if current scaling approach is good enough.

----------

I'll look into `using` tomorrow, I think I tried that and it broke but don't remember.  I know `using x = thing` is ok in classes though.

------------

> One more thing: somehow the top icons of the tab widget look messed up for me (try example1). Does it happen on your machine as well?

Do you mean the left and right arrows or the `+` for the dynamic add?

File: include/nanogui/common.h (hunk 10 above)
Comment (svenevs) [reply include/nanogui/common.h]: I'll look into `using` tomorrow, I think I tried that and it broke but don't remember.  I know `using x = thing` is ok in classes though.

(sorry if you get a lot of emails from edits, the github review thing is nice but i overcommented the first one.......)

File: src/colorpicker.cpp (hunk 24 above)
Comment (wjakob) [reply src/colorpicker.cpp]: Maybe it's me who is misunderstanding. I thought that it would be nice to have interactive feedback when e.g. changing the color of a mesh using the color picker. That requires a fast callback (i.e. "user is currently tweaking the color") and a slow one ("Pick/Reset button was finally pressed"). Potentially it will make things too complicated though.

**Comment by wjakob** [comment]
Sure, feel free to delete the other comments. Yes, I'm talking about the left and right buttons to scroll through a longer list of tabs.

**Comment by svenevs** [comment]
> Yes, I'm talking about the left and right buttons to scroll through a longer list of tabs.

I'm hoping you were only talking about **python** example 1.  The dynamic stuff wasn't in that example (see latest commit), so the icons were not displayed because there was nowhere else to scroll in the tab widget.  If you're talking about CPP this is a problem.

ColorPicker updated to have `callback` and `finalCallback` like the slider.  New window added to Example 1 to show that.

-------------------

Assuming those changes are good, the last items before this is ready are

1. Explicitly notify users about the CC-BY-SA license for Entypo+.  Proposed location is on the main README, and included in `index.rst` from the README.
2. `typedef` vs `using`.  ColorPicker fixes took longer than expected, should be able to add these two later tonight (tomorrow for you, I'm in UTC-8 right now).

**Comment by svenevs** [comment]
> 1. Explicitly notify users about the CC-BY-SA license for Entypo+. Proposed location is on the main README, and included in index.rst from the README.

See [updated `README.rst`](https://github.com/svenevs/nanogui/tree/docs-redux).  Nice thing about switching to rst is `.. contents::` works on GitHub, and now there's less duplicated text (still some though) :smile:

Please review latest commit, I moved everything to Theme because after the research I did on CC BY-SA it wasn't clear exactly how attribution should happen.  @nhojb this may or may not apply to you with ColorFinale.  In his [FAQ](http://entypo.com/faq.php) he says you could even put a comment in the source code, but that really only adheres to SA for webpages.  Not my domain...

Anyway, with all used icons being in Theme, if it is a problem then people can include a different icon font.  I added how to do that, as well as some other cmake hacks in the docs (`compilation.rst`).

> 2. typedef vs using. ColorPicker fixes took longer than expected, should be able to add these two later tonight (tomorrow for you, I'm in UTC-8 right now).

I went with `using Vector3f = Eigen::Vector3f`, but it does have a consequence.  The rendered docs give something like `using Vector3f = template Eigen::Vector3f`.  [It's been around for a while](https://github.com/michaeljones/breathe/issues/316), but `breathe` is sort of going stale.  The old-school `typedefs` work fine, and some `using` do.  I may investigate that in January.

Long story short, the current approach is good enough.  It adds build warnings and there's a weird signature, but the important part is that the docs are linkable so that when something returns a `Vector2i`, users can click on the return type and it will take them to the typedef page.  The docstrings are clear enough that it shouldn't be an issue.

**Comment by wjakob** [comment]
Great -- that sounds good, and no worries about the ``breathe`` issue, it seems fairly minor. How would you like this merged? Do you want to squash it down into a smaller number of commits?

**Comment by svenevs** [comment]
Hmm. Yeah squashing into a single commit is probably a bad idea.  I'm attaching the full [260_patch.txt](https://github.com/wjakob/nanogui/files/1319732/260_patch.txt) here.
I'll split this out into three commits: Entypo, ColorPicker, and docs.  Because I don't remember if GitHub gets mad about rebasing PRs manually, I'll create a backup branch too...

**Comment by svenevs** [comment]
Wow that was actually kind of hard.  I gave up on trying to `rebase` and just did some manual copy ops with some `git checkout HEAD -- ....` LOL.  So right now my fork has a `docs-redux-backup` branch, here's the final diff between that and the three commits here now:

```diff
diff --git a/python/example1.py b/python/example1.py
index 2bea681..237465f 100644
--- a/python/example1.py
+++ b/python/example1.py
@@ -16,7 +16,7 @@ import gc
 
 from nanogui import Color, ColorPicker, Screen, Window, GroupLayout, BoxLayout, \
                     ToolButton, Label, Button, Widget, \
-                    PopupButton, CheckBox, MessageDialog, VScrollPanel, \
+                    Popup, PopupButton, CheckBox, MessageDialog, VScrollPanel, \
                     ImagePanel, ImageView, ComboBox, ProgressBar, Slider, \
                     TextBox, ColorWheel, Graph, GridLayout, \
                     Alignment, Orientation, TabWidget, IntBox, GLShader
@@ -75,10 +75,17 @@ class TestApp(Screen):
         popup.setLayout(GroupLayout())
         Label(popup, "Arbitrary widgets can be placed here")
         CheckBox(popup, "A check box")
+        # popup right
         popupBtn = PopupButton(popup, "Recursive popup", entypo.ICON_FLASH)
-        popup = popupBtn.popup()
-        popup.setLayout(GroupLayout())
-        CheckBox(popup, "Another check box")
+        popupRight = popupBtn.popup()
+        popupRight.setLayout(GroupLayout())
+        CheckBox(popupRight, "Another check box")
+        # popup left
+        popupBtn = PopupButton(popup, "Recursive popup", entypo.ICON_FLASH)
+        popupBtn.setSide(Popup.Side.Left)
+        popupLeft = popupBtn.popup()
+        popupLeft.setLayout(GroupLayout())
+        CheckBox(popupLeft, "Another check box");
 
         window = Window(self, "Basic widgets")
         window.setPosition((200, 15))
diff --git a/src/example1.cpp b/src/example1.cpp
index d06f022..545c5c0 100644
--- a/src/example1.cpp
+++ b/src/example1.cpp
@@ -198,7 +198,6 @@ public:
         // popup left
         popupBtn = new PopupButton(popup, "Recursive popup", ENTYPO_ICON_FLASH);
         popupBtn->setSide(Popup::Side::Left);
-        popupBtn->setChevronIcon(ENTYPO_ICON_CHEVRON_LEFT);
         Popup *popupLeft = popupBtn->popup();
         popupLeft->setLayout(new GroupLayout());
         new CheckBox(popupLeft, "Another check box");
```

Thus I believe this was a success :laugh:

**Comment by wjakob** [comment]
Merged (& wiped RTD) -- thank you very much for the epic PR!

**Comment by wjakob** [comment]
Hmm -- I just noticed that this PR still breaks the tab widget. See 

https://www.dropbox.com/s/9txq8y980tomvlz/Screen%20Shot%202017-09-22%20at%2011.23.15.png?dl=0

It used to have arrows on both sides -- would you mind fixing this?

**Comment by svenevs** [comment]
Uh oh. Maybe this should be reverted. What OS are you using, Ubuntu?

**Comment by wjakob** [comment]
It's on Mac OS X.

**Comment by svenevs** [comment]
Ugh. Yeah I'm gonna have to dig a little. Main testing on my part was osx Sierra and fedora

**Comment by svenevs** [comment]
See #263 I'm really not sure what's going on here...

**Comment by svenevs** [comment]
@wjakob Eston indicated to me that the search index on RTD is somehow out of date.  E.g. if you search for `GLShader` it'll somehow link to the old one.  The only thing I can think of is re-wipe and re-launch a build.  The fake nanogui site search index is up to date, so this may just be a fluke.
